File: kattis/careful_declaration/main2.py
l1 = [i for i in input().split()]
while (l1[0] != "."):
	l2 = [i for i in input().split()]
	dp = [[0 for _ in range(len(l2)+1)] for __ in range(len(l1)+1)]
	nxt = [[(0, 0) for _ in range(len(l2)+1)] for __ in range(len(l1)+1)]
	for i in range(len(l1)+1):
		dp[i][len(l2)] = len(l1)-i
		nxt[i][len(l2)] = (i+1, len(l2))
	for i in range(len(l2)+1):
		dp[len(l1)][i] = len(l2)-i
		nxt[len(l1)][i] = (len(l1), i+1)
	for i in range(len(l1)-1, -1 -1):
		for j in range(len(l2)-1, -1, -1):
			if l1[i] == l2[j]:
				dp[i][j] = dp[i-1][j-1]+1
				nxt[i][j] = (i, j-1)
			elif dp[i-1][j] == dp[i][j-1]:
				dp[i][j] = dp[i-1][j]+1
				if l1[i-1] <= l2[j-1]:
					prev[i][j] = (i-1, j)
				else:
					prev[i][j] = (i, j-1)
			elif dp[i-1][j] < dp[i][j-1]:
				dp[i][j] = dp[i-1][j]+1
				prev[i][j] = (i-1, j)
			else:
				dp[i][j] = dp[i][j-1]+1
				prev[i][j] = (i, j-1)
	print(dp[len(l1)][len(l2)])
	i = len(l1)
	j = len(l2)
	res = []
	while i > 0 or j > 0:
		if prev[i][j] == (i-1, j-1):
			res += [l1[i-1]]
			i -= 1
			j -= 1
		elif prev[i][j] == (i-1, j):
			res += [l1[i-1]]
			i -= 1
		elif prev[i][j] == (i, j-1):
			res += [l2[j-1]]
			j -= 1
	for i in reversed(res):
		print(i, end=" ")
	print()


	'''
	res = []
	i = len(l1)
	j = len(l2)
	while i > 0 and j > 0:
		if l1[i-1] == l2[j-1]:
			res += [l1[i-1]]
			i -= 1
			j -= 1
		elif dp[i-1][j] == dp[i][j-1]:
			if l1[i-1] > l2[j-1]:
				res += [l1[i-1]]
				i -= 1
			else:
				res += [l2[j-1]]
				j -= 1
		elif dp[i-1][j] < dp[i][j-1]:
			res += [l1[i-1]]
			i -= 1
		else:
			res += [l2[j-1]]
			j -= 1
	while i > 0:
		res += [l1[i-1]]
		i -= 1
	while j > 0:
		res += [l2[j-1]]
		j -= 1
	for i in range(len(res)-1, -1, -1):
		print(res[i], end=" ")
	print()
	'''
	l1 = [i for i in input().split()]

File: ncpc/ncpc2021/custom/main2.cpp
#include <bits/stdc++.h>
using namespace std;

int n, m, k;
vector<char> res;

struct que {
	int v;
	int prev;
	int d;
	que(int v, int prev, int d) : v(v), prev(prev), d(d) {}
	bool operator > (const que& a) const {
		return d > a.d;
	}
};

void color() {
	for (int i = 0; i < n; i++)
		if (res[i] != 'x')
			continue;
		else if (k)
			k--, res[i] = 'N';
		else
			res[i] = 'S';
}
void printRes() {
	for (char c : res)
		printf("%c", c);
	printf("\n");
}

int main() {
	scanf("%d%d%d", &n, &m, &k);
	res.assign(n, 'x');
	vector<int> W(n, 0);
	for (int i = 0; i < n; i++)
		scanf("%d", &W[i]);
	vector<vector<int>> graph(n, vector<int>());
	bool edge = false;
	for (int i = 0; i < m; i++) {
		int a, b; scanf("%d%d", &a, &b); a--; b--;
		graph[a].push_back(b);
		graph[b].push_back(a);
		if ((a == 0 && b == n-1) || (a == n-1 && b == 0))
			edge = true;
	}
	if (n == 2 && k == 1) {
		printf("impossible\n");
		return 0;
	}
	else if (edge) {
		if (k >= 2)
			k -= 2, res[0] = res[n-1] = 'N';
		else
			res[0] = res[n-1] = 'S';
		color();
		printRes();
		return 0;
	}
	vector<int> toEnd;
	toEnd.push_back(n-1);
	priority_queue<que, vector<que>, greater<que>> Q;
	vector<int> dist(n, -1);
	Q.push(que(0, 0, W[0]));
	while (!Q.empty()) {
		que q = Q.top(); Q.pop();
		if (dist[q.v] == -1) {
			dist[q.v] = q.d;
			for (int nxt : graph[q.v])
				Q.push(que(nxt, q.v, q.d+W[nxt]));
		}
		if (dist[q.v] == q.d && q.v == n-1)
			toEnd.push_back(q.prev);
	}
	for (int i = 0; i < (int)toEnd.size(); i++) {
		if (!k)
			break;
		k--;
		res[toEnd[i]] = 'N';
	}
	color();
	printRes();
}
